processing/src/main/java/io/druid/segment/incremental/OffheapOakIncrementalIndex.java

```diff
@@ -0,0 +1,668 @@
+    WritableOakBufferImpl oakValue = (WritableOakBufferImpl) oak.get(serializedKey);
+    ByteBuffer aggBuffer = oakValue.getByteBuffer();
+    return agg.get(aggBuffer, aggBuffer.position() + aggOffsetInBuffer[aggPosition]);
+  }
+
+  @Override
+  protected float getMetricFloatValue(TimeAndDims timeAndDims, int aggOffset)
+  {
+    BufferAggregator agg = getAggs()[aggOffset];
+    ByteBuffer serializedKey = timeAndDimsSerialization(timeAndDims);
+    WritableOakBufferImpl oakValue = (WritableOakBufferImpl) oak.get(serializedKey);
+    ByteBuffer aggBuffer = oakValue.getByteBuffer();
+    return agg.getFloat(aggBuffer, aggBuffer.position() + aggOffsetInBuffer[aggOffset]);
+  }
+
+  @Override
+  protected long getMetricLongValue(TimeAndDims timeAndDims, int aggOffset)
+  {
+    BufferAggregator agg = getAggs()[aggOffset];
+    ByteBuffer serializedKey = timeAndDimsSerialization(timeAndDims);
+    WritableOakBufferImpl oakValue = (WritableOakBufferImpl) oak.get(serializedKey);
+    ByteBuffer aggBuffer = oakValue.getByteBuffer();
+    return agg.getLong(aggBuffer, aggBuffer.position() + aggOffsetInBuffer[aggOffset]);
+  }
+
+  @Override
+  protected Object getMetricObjectValue(TimeAndDims timeAndDims, int aggOffset)
+  {
+    BufferAggregator agg = getAggs()[aggOffset];
+    ByteBuffer serializedKey = timeAndDimsSerialization(timeAndDims);
+    WritableOakBufferImpl oakValue = (WritableOakBufferImpl) oak.get(serializedKey);
+    ByteBuffer aggBuffer = oakValue.getByteBuffer();
+    return agg.get(aggBuffer, aggBuffer.position() + aggOffsetInBuffer[aggOffset]);
+  }
+
+  @Override
+  protected double getMetricDoubleValue(TimeAndDims timeAndDims, int aggOffset)
+  {
+    BufferAggregator agg = getAggs()[aggOffset];
+    ByteBuffer serializedKey = timeAndDimsSerialization(timeAndDims);
+    WritableOakBufferImpl oakValue = (WritableOakBufferImpl) oak.get(serializedKey);
+    ByteBuffer aggBuffer = oakValue.getByteBuffer();
+    return agg.getDouble(aggBuffer, aggBuffer.position() + aggOffsetInBuffer[aggOffset]);
+  }
+
+  @Override
+  protected boolean isNull(TimeAndDims timeAndDims, int aggOffset)
+  {
+    BufferAggregator agg = getAggs()[aggOffset];
+    ByteBuffer serializedKey = timeAndDimsSerialization(timeAndDims);
+    WritableOakBufferImpl oakValue = (WritableOakBufferImpl) oak.get(serializedKey);
+    ByteBuffer aggBuffer = oakValue.getByteBuffer();
+    return agg.isNull(aggBuffer, aggBuffer.position() + aggOffsetInBuffer[aggOffset]);
+  }
+
+  @Override
+  public Iterable<TimeAndDims> timeRangeIterable(
+      boolean descending, long timeStart, long timeEnd)
+  {
+    if (timeStart > timeEnd) {
+      return null;
+    }
+    ByteBuffer from = timeAndDimsSerialization(new TimeAndDims(timeStart, null, dimensionDescsList));
+    ByteBuffer to = timeAndDimsSerialization(new TimeAndDims(timeEnd + 1, null, dimensionDescsList));
+    OakMap subMap = oak.subMap(from, true, to, false);
+    if (descending == true) {
```

Comment:
if (descending)

Reply:
Hi @jihoonson ,
Thanks for taking a look! The pull request is really an initial one and will be improved before we ask for a merge. Your comments are very welcome!
Regarding the open sourcing the code of Oak and its license, we currently discussing it internally and we will update you soon.

@kaijianding ,
Thank you as well, we will fix those styling issues.

Reply:
@sanastas great! Looking forward to your contribution!